src/signalrclient/connection_impl.cpp

```diff
@@ -623,10 +588,75 @@ namespace signalr
             change_state(connection_state::disconnecting);
         }
 
-        m_transport->stop([callback](std::exception_ptr exception)
+        if (m_transport)
+        {
+            m_transport->stop(callback);
+        }
+        else
+        {
+            stop_connection(nullptr);
+        }
+    }
+
+    // do not use `shared_from_this` as it can be called via the destructor
+    void connection_impl::stop_connection(std::exception_ptr error)
+    {
+        {
+            // the lock prevents a race where the user calls `stop` on a disconnected connection and calls `start`
+            // on a different thread at the same time. In this case we must not null out the transport if we are
+            // not in the `disconnecting` state to not affect the 'start' invocation.
+            std::lock_guard<std::mutex> lock(m_stop_lock);
+
+            if (m_connection_state == connection_state::connecting)
             {
-                callback(exception);
-            });
+                // should never happen
+                m_logger.log(trace_level::errors, "Stopping was ignored because the connection is still in the connecting state.");
```

Comment:
Couldn't this happen if the user tries to stop the connection before start() completes

If this really cannot happen, this should probably throw just in case. If this can happen, this should cancel start() and complete after start() fails or after start() completes and then the connection disconnects.

Reply:
No, because stop will wait for start to finish before calling this function.

However, in the current state there is a race where the transport could connect and then close due to an error before the state is changed to connected which would hit this code path I believe.